File: QuizGames/QuizGames.AspNetCore/wwwroot/js/gameHub.js
//
// Variables

// Custom event constant
const scoreUpdated =
    {
        Dispatch: function (playerId, score) {
            this.handlers.forEach((func) => {
                func(playerId, score);
            });
        },
        handlers: []
    };

const playerJoined =
    {
        Dispatch: function (users) {
            this.handlers.forEach((func) => {
                func(users);
            });
        },
        handlers: []
    };

const gameStarted =
    {
        Dispatch: function () {
            this.handlers.forEach((func) => {
                func();
            });
        },
        handlers: []
    };

const clientConnected =
    {
        Dispatch: function () {
            this.handlers.forEach((func) => {
                func();
            });
        },
        handlers: []
    };

// Contains the connected players & your player name
var players = new Array();

//
// Create connection to game hub
let connection = new signalR.HubConnectionBuilder()
    .withUrl("/hub/game")
    .configureLogging(signalR.LogLevel.Information)
    .build();

//
// Window close event (prevents players from taking up player space when they left the game)
window.onbeforeunload = function () {
    connection.invoke("LeaveGame");
};


//
// Add connection handlers

// Client attempted to start the game
connection.on("GameStarted", function (isStartable) {
    if (isStartable)
        gameStarted.Dispatch();
});

// Client attempted to connect to the game
connection.on("GameConnected", function (playerClients) {
    if (playerClients === -1) {
        // Alert and redirect to previous page
        alert("The game is already full!");
        history.go(-1);
    }
    else if (playerClients === -2)
    {
        // Alert and redirect to previous page
        alert("You're already in the game!");
        history.go(-1);
    }
    else if (playerClients === -3) {
        // Alert and redirect to previous page
        alert("The game has already started!");
        history.go(-1);
    }
    else {
        // Get local username
        let playerName = getCookie("playerName");

        // Get local user and remaining users
        let locPlayer = playerClients.find((player) => { return player.playerId === playerName; });
        let remPlayers = Array();

        // Add previously connected clients to player list
        for (let i = 0; i < playerClients.length; i++) {
            // Add remaining players to player list
            if (playerClients[i].playerId !== playerName)
                remPlayers[i] = playerClients[i];
        }

        // Set the player list
        players = remPlayers;
        players.unshift(locPlayer);

        // Notify console
        console.log("Connected!");
        clientConnected.Dispatch();
    }
});

// A player connected to the game
connection.on("PlayerConnected", function (player) {
    // Add player to the player list
    players[players.length] = player;
    playerJoined.Dispatch(players);

    // Send the clients current score to the newly connected player
    connection.invoke("SendScore", players[0].score);
});

// A player disconnected from the game
connection.on("PlayerDisconnected", function (player) {
    // Find player in the player list
    let index = players.findIndex(function (item) {
        return item.playerId === player.playerId;
    });

    // Remove the player from the list
    players.splice(index, 1);
});

// Change a player's score
connection.on("ReceiveScore", function (player, score) {
    // Find player in the player list
    let index = players.findIndex(function (item) {
        return item.playerId === player.playerId;
    });

    // Assign new score
    players[index].score = score;

    // Dispatch score update event
    scoreUpdated.Dispatch(player.playerId, score);
});

//
// Start connection and attempt to join the game
connection.start()
    .then(function () {
        connection.invoke("JoinGame", getCookie("playerName"));
    })
    .catch(function (error) {
        console.error(error.toString());
    });

File: QuizGames/QuizGames.AspNetCore/Hubs/GameHub.cs
using Microsoft.AspNetCore.SignalR;
using QuizGames.ViewModels;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace QuizGames.AspNetCore.Hubs
{
    public class GameHub : Hub
    {
        // Contains a global count of players (game management)
        public static List<GameViewModel> players;
        public static bool gameState = false;

        // A client starts the game
        public async Task StartGame()
        {
            await Clients.All.SendAsync("GameStarted", players.Count == 4);

            // Started game
            gameState = true;
        }

        // A client attempts to join the game
        public async Task JoinGame(string playerId)
        {
            // Return if game started
            if (gameState)
            {
                await Clients.Caller.SendAsync("GameConnected", -3);
                return;
            }

            if (players == null || players.Count < 4)
            {
                // Check if the player is already in game
                if (players != null && players.Count((item) => { return item.PlayerId == playerId; }) >= 1)
                {
                    await Clients.Caller.SendAsync("GameConnected", -2);
                    return;
                }

                // Create player
                GameViewModel player = new GameViewModel(Context.ConnectionId, playerId);

                // Create a player list (if null) add the new connection
                players = players ?? new List<GameViewModel>();
                players.Add(player);

                // Send the connection to the other players and confirm connection and retrieve currently connected players
                await Clients.Others.SendAsync("PlayerConnected", player);
                await Clients.Caller.SendAsync("GameConnected", players);
            }
            else
            {
                // Affirm connection
                await Clients.Caller.SendAsync("GameConnected", -1);
            }
        }

        // A client attempts to leave the game
        public async Task LeaveGame()
        {
            if (players.Count(item => { return item.ConnectionId == Context.ConnectionId; }) >= 1)
            {
                // Get player
                GameViewModel player = players.First(item => { return item.ConnectionId == Context.ConnectionId; });

                // Remove the player from the list
                players.Remove(player);

                // Reset gamestate if all players have left
                gameState = gameState && players.Count <= 0 ? false : gameState;

                // Reset the player score
                await Clients.Others.SendAsync("ReceiveScore", player, 0);

                // Send disconnection to other players and confirm disconnection
                await Clients.Others.SendAsync("PlayerDisconnected", player);
                await Clients.Caller.SendAsync("GameDisconnected", true);
            }
            else
            {
                // Affirm disconnection (can't disconnect when not connected)
                await Clients.Caller.SendAsync("GameDisconnected", false);
            }
        }

        // A player attempts to update their (online) score
        public async Task SendScore(int score)
        {
            // Send score to other clients
            await Clients.Others.SendAsync("ReceiveScore", players.First((item) => { return item.ConnectionId == Context.ConnectionId; }), score);
        }
    }
}
